### DISCUSSION TITLE
Weaken dependency constraints

### DISCUSSION THREAD

**Discussion by pathob** [make-all/tuya-local#1998] [General]
Hi @make-all,

[I've made tuya-local available for NixOS](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=tuya_local) with version 2024.5.2 and have issues upgrading it to version 2024.6.0 because of the hard dependency constraints tuya-local has.

tuya-local since a few weeks requires tuya-device-sharing-sdk 0.1.9, which was not the latest version (0.2.0) when added to tuya-local. [In NixOS the packaged version of tuya-device-sharing-sdk is "already" 0.2.0.](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=tuya-device-sharing-sdk)

[And I've also packaged tinytuya for NixOS](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=tinytuya), currently with version 1.13.2.

The "issue" with NixOS is, that every package must be 100% reproducible and thus all Python dependencies of a NixOS package must be packaged in NixOS and be 100% reproducible itself also. With the current version constraints, I'm not able to upgrade any of the packages anymore, because upgrading tinytuya would break tuya-local (still with version 2024.5.2) and upgrading tinytuya and tuya-local together would stilll break tuya-local because of the version mismatch with the tuya-device-sharing-sdk.

The easiest solution for this problem would be to weaken the dependency constraints of tuya-local. Assuming that the dependencies properly implement semantic versioning, they wouldn't have any breaking changes unless the major version changes - but only receive new features with minor releases or bugfixes with patch releases. And under this assumption, the current version constraints could maybe changed from

```
tinytuya~=1.14.0
tuya-device-sharing-sdk==0.1.9
```

to

```
tinytuya>=1.14.0
tuya-device-sharing-sdk>=0.1.9
```

assuming that these dependency versions have been chosen to contain a required feature (tinytuya) respectively a required bugfix (tuya-device-sharing-sdk). Of course other version specifiers are also possible, but it really depends on what features or bugfixes the project _actually_ requires.

Please let me know what you think about this. And I could also provide a corresponding merge request.

**Comment by make-all** [comment]
This would bring risk of bugs/regression if the library was updated incompatibly.  Using ~= ensures that major version releases are not automatically upgraded, allowing them to be tested first.

I don't know what NixOS is, and why it pre-packages Home Assistant integrations like this, normally users use HACS.  Last time I checked (a few days ago) Home Assistant was also using tuya-sharing-sdk 0.1.9, and the most important thing it to remain compatible with that, though if 0.2.0 is now released, it can be tested and included in the compatible versions somehow.

**Comment by make-all** [comment]
Version 0.2.0 seems to be the unreleased HEAD of dev branch. I do not think it is appropriate to use for this integration now.

**Reply by pathob** [reply]
Thank you for your assessment. At first I was considering that version 0.2.0 as a regular release that got released from a wrong branch, probably because the maintainer doesn't know better, but I can follow your reasoning too. And since the official Tuya integration also sticks to version 0.1.9, we've decided at NixOS to downgrade the library, so that I was now able to build tuya-local even with the "hard" dependency constraints.

Just a thing that I would like to ask you. Please give other people the possibility to respond to your hesitations before you close their issues.

Thank you
